src/guess_density.jl

```diff
@@ -1,59 +1,70 @@
+struct AutoGuessDensity <: GuessDensityMethod end
 
 @doc raw"""
-    guess_density(basis, magnetic_moments=[])
-    guess_density(basis, system)
+    guess_density(basis::PlaneWaveBasis, method::GuessDensityMethod,
+                  magnetic_moments=[])
 
 Build a superposition of atomic densities (SAD) guess density.
 
-We take for the guess density a Gaussian centered around the atom, of
-length specified by `atom_decay_length`, normalized to get the right number of electrons
+The guess atomic densities are taken as one of the following depending on the input
+`method`:
+
+-`RandomGuessDensity(n_electrons)`: A random density, normalized to the number of electrons
+provided. Does not support magnetic moments.
+-`AutoGuessDensity()`: A combination of the `:gaussian` and `:psp` methods where
+elements whose pseudopotentials provide numeric valence charge density data use them and
+elements without use Gaussians.
+-`GaussianGuessDensity()`: Gaussians of length specified by `atom_decay_length`
+normalized for the correct number of electrons:
 ```math
 \hat{ρ}(G) = Z \exp\left(-(2π \text{length} |G|)^2\right)
 ```
+- `PspGuessDensity()`: Numerical pseudo-atomic valence charge densities from the
+pseudopotentials. Will fail if one or more elements in the system has a pseudopotential
+that does not have valence charge density data.
+
 When magnetic moments are provided, construct a symmetry-broken density guess.
 The magnetic moments should be specified in units of ``μ_B``.
 """
-function guess_density(basis::PlaneWaveBasis, magnetic_moments=[])
-    guess_density(basis, basis.model.atoms, basis.model.positions, magnetic_moments)
-end
-function guess_density(basis::PlaneWaveBasis, system::AbstractSystem)
-    parsed = parse_system(system)
-    guess_density(basis, parsed.atoms, parsed.positions, parsed.magnetic_moments)
-end
-@timing function guess_density(basis::PlaneWaveBasis, atoms, positions, magnetic_moments)
-    ρtot = _guess_total_density(basis, atoms, positions)
-    if basis.model.n_spin_components == 1
-        ρspin = nothing
-    else
-        ρspin = _guess_spin_density(basis, atoms, positions, magnetic_moments)
+function guess_density(basis::PlaneWaveBasis{T}, method::GuessDensityMethod,
+                       magnetic_moments=[]) where {T}
+    ρtot = _atomic_density_superposition(basis, method)
+    ρspin = _guess_spin_density(basis, method, magnetic_moments)
+    ρ = ρ_from_total_and_spin(ρtot, ρspin)
+    
+    Z = sum(ρ) * basis.model.unit_cell_volume / prod(basis.fft_size)
+
+    isnothing(basis.model.n_electrons) && return ρ
+
+    if (abs(Z - basis.model.n_electrons) > sqrt(eps(eltype(ρ)))) &
+        (basis.model.n_electrons > 0) & (Z > 0)
+        ρ *= basis.model.n_electrons / Z  # Renormalize to the correct number of electrons
+        Z_renormalized = sum(ρ) * basis.model.unit_cell_volume / prod(basis.fft_size)
+        @info @sprintf "Renormalized guess density from %0.9f to %0.9f" Z Z_renormalized
```

Comment:
remove debug statement?

Reply:
I wasn't really using it as a debug statement, QE prints it as well and it seemed a bit useful. I guess at the end of the day you don't _have_ to know about this, so I can remove it.

As a side note, it would be nice if DFTK had a logging infrastructure where messages like this could go.